wrangler-core/src/main/java/io/cdap/wrangler/utils/ArithmeticOperations.java

```diff
@@ -388,12 +391,64 @@ public static Float min(Float f1, Float f2) {
   }
 
   /**
-   Arithmetic operation - Find the minimum of two columns.
+   * Arithmetic operation - Find the minimum of two columns.
    */
   public static BigDecimal min(BigDecimal bd1, BigDecimal bd2) {
     if (bd1 == null || bd2 == null) {
       return null;
     }
     return bd1.min(bd2);
   }
+
+  /**
+   * Arithmetic operation - calculate average of any number of columns.
+   * Returns Number - BigDecimal if any column is of type BigDecimal, double otherwise.
+   * HALF_EVEN rounding mode chosen for high precision in the long run.
+   */
+  public static Number average(Number... nums) throws DirectiveExecutionException {
+    // Note that this algorithm avoids overflow
+    Number avg = 0.0;
+    int t = 1;
+    boolean allNull = true;
+    boolean noBigDecimal = true;
+    for (Number num : nums) {
+      // Check for null input
+      if (num == null) {
+        continue;
+      } else {
+        allNull = false;
+      }
+      // Check for invalid input
+      if (num instanceof Byte) {
+        throw new DirectiveExecutionException("Input cannot be of type 'Byte'.");
+      }
+      // Compute average (no BigDecimal case)
+      if (noBigDecimal && num instanceof BigDecimal) {
```

Comment:
This logic is a bit confusing.

As far as I see it, either we are doing BigDecimal arithmetics or we are not. As soon as you find one BigDecimal you switch into "BigDecimal mode" and then all calculations are performed over BigDecimals.

I would reorder this as:

1. have a variable that tracks if you have found a BigDecimal (similar to noBigDecimal)
2. Check if the current value is an instance of BigDecimal. If true, go into BigDecimal arithmetics mode (more precise). if false, use double arithmetics (faster).

I think it's also important to point out that the output schema is going to be different in each case. As soon as you enter "BigDecimal" mode, the output should be a decimal. If not, it should be double.

Reply:
> As far as I see it, either we are doing BigDecimal arithmetics or we are not. As soon as you find one BigDecimal you switch into "BigDecimal mode" and then all calculations are performed over BigDecimals.
Yes, that's the idea

I'm not quite sure what you are saying -- are you recommending that we create two other functions (one for BigDecimal, one for double) to use depending on the situation?

Reply:
The logic should now be more clear after Ellen's most recent updates.